coucharchive

```diff
@@ -342,13 +374,14 @@ def replicate_couchdb_server(source_url, target_url,
 
     in_queue = multiprocessing.Queue()
     out_queue = multiprocessing.Queue()
-    control = ReplicationControl(len(dbs), ideal_duration)
+    control = ReplicationControl(len(dbs), ideal_duration, source_url,
+                                 target_url, has_run_queue_control)
     pool = multiprocessing.Pool(MAX_NUMBER_OF_WORKERS, replicate_databases,
                                 (in_queue, out_queue, control, source_url,
                                  target_url, reuse_db_if_exist))
     error = None
 
-    while len(dbs) and not error:
+    while len(dbs) and not error and control.is_alive():
```

Comment:
If I understand correctly:
1. If the run_queue thread dies (e.g. random network problem), it will stop coucharchive.
2. If the run_queue thread dies, ir won't raise any error nor bad exit code: the user won't know that replication was stopped in the middle.

Reply:
The error trace is displayed by the thread itself.
So it doesn't return an error like replications process, but it's displayed in `stderr` like other traces of `coucharchive`.
For example, raising a fake `toto` exception in the thread give:
```python trace
INFO: blabla: done
INFO: Oo: done
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib64/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "./coucharchive", line 273, in _run_queue_control_function
    raise Exception('toto')
Exception: toto

INFO: Terminating local CouchDB instance
INFO: Creating backup archive at toto.tar.gz
```

Reply:
OK, so we have a problem, no? If the backup file is created and no error code returned when the control thread dies and the replication loop stopped, the user will think it's OK while it's not? (by the way, the traceback will be way up in logs, if you have 256 workers)

Reply:
Hey Adrien!
You are right I haven't been carrefull with error return code of the program, which is wrong in this case.
I have too solution to propose you:
1. https://github.com/adrienverge/coucharchive/commit/ca1b93a8c5f4d424bd2539c29506ff5ac39d5707
2. https://github.com/adrienverge/coucharchive/commit/42e9072516378e2be765420060f7bb6c69815bf5

I let you tell me if you like one.
On my part, over the two, I prefer the second, because it re-use the `getattr(t, "do_run", True)` to store the exception, and thus handle the issue you mentioned in bracket (`by the way, the traceback will be way up in logs, if you have 256 workers`)

Reply:
Subject discussed in private messages, solution 1. chose!
```diff
--- a/coucharchive
+++ b/coucharchive
@@ -396,7 +396,7 @@ def replicate_couchdb_server(source_url, target_url,
                                  target_url, reuse_db_if_exist))
     error = None
 
-    while len(dbs) and not error and control.is_alive():
+    while len(dbs) and not error:
         ideal = control.ideal_number_of_replications()
         while len(dbs) and control.running_replications.value < ideal:
             in_queue.put(dbs.pop(0))
@@ -415,6 +415,10 @@ def replicate_couchdb_server(source_url, target_url,
 
         time.sleep(1)
 
+        if not control.is_alive():
+            error = Exception('Control died.')
+            break
+
         while True:
             try:
                 result = out_queue.get_nowait()